README.md
# 微信跳一跳游戏助手

 [GitHub 项目地址](https://github.com/xushanmeng/WechatJumpHelper)
 
 微信跳一跳技术讨论QQ群148698890
 
 [更新日志](changelog.md)
 
 Windows一键开始请移步[release](release)
 
 用户常见问题请先查询[FAQ](FAQ.md)

## 功能简介

>用JAVA自动控制手机玩跳一跳

* 自动识别图像计算距离
* 自动帮你点击屏幕
* 自动缓存图片，并在图片上标记一些识别结果，如下图![示例图片](Samples/172_mark.png)


## 运行环境
1. JAVA，最低版本为7.0，[官网下载](http://www.oracle.com/technetwork/java/javase/downloads/index.html)
2. 安卓手机，目前已适配分辨率
    * 1600x2560
    * 1440x2560
    * 1080x1920
    * 720x1280

## 使用方法

>有JAVA开发工具的同学可以直接运行java代码，便于代码调试，下面主要介绍运行已经打包好的jar包的方法

1. 手机打开USB调试，并连接电脑
    * 打开USB调试方法，进入`设置`，找到`开发者选项`，打开并勾选`USB调试`；
    * 如果没有`开发者选项`，进入`关于手机`，连续点击`版本号`7次，即可开启`开发者选项`。   
2. 通过下面的命令，运行[Android.jar](Android/build/libs/Android.jar)
    ```
    java -jar Android.jar
    ```   
3. 根据手机分辨率选择跳跃系数，目前已适配机型：

    * 1600x2560机型推荐0.92
    * 1440x2560机型推荐1.039
    * 1080x1920机型推荐1.392
    * 720x1280机型推荐2.078 
    
    其他分辨率请自己微调。
    
## 原理说明
1. 通过adb命令控制手机截图，并取回到本地

    ```
    adb shell screencap -p /sdcard/screen.png
    adb pull /sdcard/screen.png .
    ```
    
2. 图片分析
    * 根据棋子的颜色，取顶部和底部的特征像素点，在截图中进行匹配，找到棋子坐标
    * 由于目标物体不是在左上就是在右上，可以从上往下扫描，根据色差判断目标物体位置，其中又分为以下几种类型
        * 有靶点，即目标物体中心的白色圆点，则靶点中心为目标落点
        
        ![](Samples/412_mark.png)
        * 无靶点，但是纯色平面，或者规则平面，则平面中心为目标落点
        
        ![](Samples/444_mark.png)
        * 无靶点，又无纯色规则平面，但是左上和右上位置的斜率是固定的，可根据固定斜率的斜线和目标物体中心线的焦点计算落点
        
        ![](Samples/544_mark.png)
    * 计算棋子坐标和目标落点的距离
    * 距离×跳跃系数=按压屏幕的时间，不同分辨率的手机，跳跃系数也有所不同
    
3. 通过adb命令，给手机模拟按压事件
    ```
    adb shell input swipe x y x y time
    ```
    其中`x`和`y`是屏幕坐标，`time`是触摸时间，单位ms。
    
## 工程结构
```目录结构
.
├── Samples             # 一些示例图片
├── Android             # 使用Android手机的相关代码和jar包
│   ├── src             # JAVA源代码目录
│   └── build           # Android.jar包所在目录
└── README.md
```
## 代码详解

>这里将针对一些关键算法的代码进行解释

#### 1. 寻找棋子位置

把截图放大，可以看到棋子顶部像素连成一条横线，那么我们通过颜色匹配，找到这一条线的始末位置，取中间位置，就得到了棋子的x坐标。

![](Samples/piece_top.png)

棋子的底部也是一条横线，用颜色匹配，我们检测到相似颜色的最大y坐标，就是棋子底部了，不过考虑到棋子底部是个圆盘，我们把棋子的y坐标再往上提一些。

![](Samples/piece_bottom.png)

这样我们就得到了棋子的xy坐标，下面是相关代码：

```java
/* 计算棋子位置 */
Pixel piece = new Pixel();
for (int i = TOP_BORDER; i < screenHeight - BOTTOM_BORDER; i++) {
    int startX = 0;
    int endX = 0;
    for (int j = LEFT_BORDER; j < screenWidth - RIGHT_BORDER; j++) {
        int red = Color.red(pixels[i][j].color);
        int green = Color.green(pixels[i][j].color);
        int blue = Color.blue(pixels[i][j].color);
        if (50 < red && red < 55
                && 50 < green && green < 55
                && 55 < blue && blue < 65) {//棋子顶部颜色
            //如果侦测到棋子相似颜色，记录下开始点
            if (startX == 0) {
                startX = j;
                endX = 0;
            }
        } else if (endX == 0) {
            //记录下结束点
            endX = j;

            if (endX - startX < PIECE_TOP_PIXELS) {
                //规避井盖的BUG，像素点不够长，则重新计算
                startX = 0;
                endX = 0;
            }
        }
        if (50 < red && red < 60
                && 55 < green && green < 65
                && 95 < blue && blue < 105) {//棋子底部的颜色
            //最后探测到的颜色就是棋子的底部像素
            piece.y = i;
        }
    }
    if (startX != 0 && piece.x == 0) {
        piece.x = (startX + endX) / 2;
    }
}
//棋子纵坐标从底部边缘调整到底部中心
piece.y -= PIECE_BOTTOM_CENTER_SHIFT;
```

#### 2. 寻找靶点

所谓靶点，就是目标物体中心的那个小圆点，颜色值为`0xf5f5f5`。

![](Samples/target.png)

那么我们只需要寻找颜色值为0xf5f5f5的色块就可以了，为了规避其他物体相近颜色干扰，我们可以限制色块的大小，正确大小的色块才是靶点。

但是如何计算色块的大小呢，色块最顶端到最底端y坐标的差值我们作为色块的高度，同理，最左侧到最右侧x坐标的差值作为宽度，我们只需要查找这四个顶点的坐标就可以了。

本来打算用凸包的Graham扫描算法，后来发现色块已经是凸包了，且边缘像素是连续的，那么我们按照一定顺序，遍历边缘像素，就可以在O(n^-2)的时间复杂度里，得到色块的顶点坐标了。

我们从第一个像素点开始，寻找的顺序如图所示：

![](Samples/find_vertex.png)

```java
    /**
     * 寻找色块顶点像素
     */
    public static final Pixel[] findVertexs(Pixel[][] pixels, Pixel firstPixcel) {
        Pixel[] vertexs = new Pixel[4];
        Pixel topPixel = firstPixcel;
        Pixel leftPixel = firstPixcel;
        Pixel rightPixel = firstPixcel;
        Pixel bottomPixel = firstPixcel;
        Pixel currentPixcel = firstPixcel;
        //先把坐标置于左上角
        while (checkBorder(pixels, currentPixcel)//判断是否超出图像边缘
                && Color.compareColor(pixels[currentPixcel.y - 1][currentPixcel.x], firstPixcel)) {//判断是否是相同颜色
            currentPixcel = pixels[currentPixcel.y - 1][currentPixcel.x];
        }
        while (checkBorder(pixels, currentPixcel)
                && Color.compareColor(pixels[currentPixcel.y][currentPixcel.x - 1], firstPixcel)) {
            currentPixcel = pixels[currentPixcel.y][currentPixcel.x - 1];
        }
        //寻找上顶点像素
        while (checkBorder(pixels, currentPixcel)) {
            if (Color.compareColor(pixels[currentPixcel.y - 1][currentPixcel.x], firstPixcel)) {
                currentPixcel = pixels[currentPixcel.y - 1][currentPixcel.x];
            } else if (Color.compareColor(pixels[currentPixcel.y][currentPixcel.x + 1], firstPixcel)) {
                currentPixcel = pixels[currentPixcel.y][currentPixcel.x + 1];
            } else {
                topPixel = findCenterPixcelHorizontal(pixels, currentPixcel);
                break;
            }
        }
        //寻找右顶点像素
        while (checkBorder(pixels, currentPixcel)) {
            if (Color.compareColor(pixels[currentPixcel.y][currentPixcel.x + 1], firstPixcel)) {
                currentPixcel = pixels[currentPixcel.y][currentPixcel.x + 1];
            } else if (Color.compareColor(pixels[currentPixcel.y + 1][currentPixcel.x], firstPixcel)) {
                currentPixcel = pixels[currentPixcel.y + 1][currentPixcel.x];
            } else {
                rightPixel = findCenterPixcelVertial(pixels, currentPixcel);
                break;
            }
        }
        //寻找下顶点像素
        while (checkBorder(pixels, currentPixcel)) {
            if (Color.compareColor(pixels[currentPixcel.y + 1][currentPixcel.x], firstPixcel)) {
                currentPixcel = pixels[currentPixcel.y + 1][currentPixcel.x];
            } else if (Color.compareColor(pixels[currentPixcel.y][currentPixcel.x - 1], firstPixcel)) {
                currentPixcel = pixels[currentPixcel.y][currentPixcel.x - 1];
            } else {
                bottomPixel = findCenterPixcelHorizontal(pixels, currentPixcel);
                break;
            }
        }
        //寻找左顶点像素
        while (checkBorder(pixels, currentPixcel)) {
            if (Color.compareColor(pixels[currentPixcel.y][currentPixcel.x - 1], firstPixcel)) {
                currentPixcel = pixels[currentPixcel.y][currentPixcel.x - 1];
            } else if (Color.compareColor(pixels[currentPixcel.y - 1][currentPixcel.x], firstPixcel)) {
                currentPixcel = pixels[currentPixcel.y - 1][currentPixcel.x];
            } else {
                leftPixel = findCenterPixcelVertial(pixels, currentPixcel);
                break;
            }
        }
        vertexs[0] = leftPixel;
        vertexs[1] = topPixel;
        vertexs[2] = rightPixel;
        vertexs[3] = bottomPixel;
        return vertexs;
    }
```
得到了四个坐标点，我们就可以计算色块的中点了，也就是目标落点。

对于没有靶点，但是落点是规则平面的，也可以用类似算法。

#### 3. 斜率计算

对于没有靶点，又不是规则平面的，我们怎么计算落点呢，这时候就要用到斜率了。

可以看得出来，每次左上角或右上角出现的物体，针对当前物体的方向都是一样的，也就是两个物体中心的连线，斜率是固定的。

基本所有的目标物体，最顶点像素中点的x坐标，都是在物体中间，我们至少先得到了目标物体x坐标了，记为des.x ，接下来要求des.y 。

![](Samples/k1.png)

如上图所示，计算过程如下：
```
    斜线的公式为 y=kx+b 
    那么，在棋子坐标上有 piece.y=k*piece.x+b 
    在目标落点坐标上有 des.y=k*des.x+b
    代入得到 des.y=k*(des.x-piece.x)+piece.y
```
然而这种算法还是有偏差的。

![](Samples/k2.png)

可以看到，同样的斜率，如果棋子的位置有偏差，计算出来最终落点还是会有偏差的。

代码解析就先讲这么多，希望有大神可以提出更好的解决方案。
 
## 玩游戏小窍门
1. 连续的落到物体中心位置，是有分数加成的，最多跳一次可以得几十分
2. 井盖、商店、唱片、魔方，多停留一会，有音乐响起后也是有分数加成的
>那么看一下程序员的朋友圈有多残酷吧

![](Samples/rank.jpeg)


changelog.md
## 更新日志

* 2018-01-05:
    * 集成了Windows和Mac的adb，可以不用另装adb运行了
    * 添加了手机连接的检测
    * 修复了全面屏手机计算失败的BUG

* 2018-01-04:
    * 调整了分辨率检测策略，用户可以自定义跳跃系数
    * 模拟点击位置改为随机位置，修复了因为微信外挂检测而无法上传积分的问题

* 2018-01-03:
    * 添加了手机分辨率的自动检测
    * 修正了720x1280分辨率的问题

* 2018-01-02:
    * 初始提交

Android/src/com/tencent/wechatjump/helper/Helper.java
package com.tencent.wechatjump.helper;

import com.tencent.wechatjump.helper.bean.DesType;
import com.tencent.wechatjump.helper.bean.Pixel;
import com.tencent.wechatjump.helper.util.Color;
import com.tencent.wechatjump.helper.util.DesTypeChecker;
import com.tencent.wechatjump.helper.util.FileUtil;
import com.tencent.wechatjump.helper.util.HelperUtil;

import java.awt.Font;
import java.awt.Graphics;
import java.awt.image.BufferedImage;
import java.io.File;

import javax.imageio.ImageIO;

import static com.tencent.wechatjump.helper.util.DebugUtil.*;

/**
 * Created by xushanmeng on 2018/1/2.
 */

public class Helper {

    private double jumpParam;
    private StringBuilder debugInfo;
    private int cacheIndex = 0;
    private File cacheDir = new File(Constants.BASE_DIR, "imageCaches");
    private File markDir = new File(Constants.BASE_DIR, "imageMarks");

    public Helper(double jumpParam) {
        this.jumpParam = jumpParam;
    }

    public void start() {
        if (!DEBUG) {
            if (!cacheDir.exists()) {
                cacheDir.mkdirs();
            }
            if (!markDir.exists()) {
                markDir.mkdirs();
            }
            String[] filenames = cacheDir.list();
            for (String filename : filenames) {
                String[] seg = filename.split("\\.");
                try {
                    int index = Integer.valueOf(seg[0]);
                    if (index > cacheIndex) {
                        cacheIndex = index;
                    }
                } catch (Exception e) {
                    continue;
                }
            }
        }
        while (true) {
            try {
                debugInfo = new StringBuilder();
                long startTime = System.currentTimeMillis();
                long screencapTime = 0;
                long downloadTime = 0;
                long calculateTime = 0;
                long cacheTime = 0;
                long exeJumpTime = 0;
                long recacheTime = 0;
                long tempTime = startTime;
                File cacheFile;
                if (!DEBUG) {
                    // 获取截图
                    boolean result = HelperUtil.execute(Constants.ADB_PATH.getAbsolutePath(), "shell", "screencap", "-p", "/sdcard/screen.png");
                    if (!result) {
                        System.out.println("截屏失败");
                        Thread.sleep(5000);
                        continue;
                    } else {
                        screencapTime = System.currentTimeMillis() - tempTime;
                    }
                    tempTime += screencapTime;
                    result = HelperUtil.execute(Constants.ADB_PATH.getAbsolutePath(), "pull", "/sdcard/screen.png", Constants.BASE_DIR.getAbsolutePath());
                    if (!result) {
                        System.out.println("下载截屏失败");
                        Thread.sleep(5000);
                        continue;
                    } else {
                        downloadTime = System.currentTimeMillis() - tempTime;
                    }
                    tempTime += downloadTime;
                    cacheFile = new File(cacheDir, (++cacheIndex) + ".png");
                    if (cacheIndex > Constants.CACHE_FILE_MAX) {
                        // 最多缓存1000张图片
                        File oldFile = new File(cacheDir, (cacheIndex - Constants.CACHE_FILE_MAX) + ".png");
                        oldFile.delete();
                    }
                    // 保存缓存图片
                    File imageFile = new File(Constants.BASE_DIR, "screen.png");
                    result = FileUtil.copyFile(imageFile, cacheFile);
                    imageFile.deleteOnExit();
                    if (!result) {
                        System.out.println(cacheIndex + ". 截屏缓存保存失败");
                    }
                    cacheTime = System.currentTimeMillis() - tempTime;
                    tempTime += cacheTime;
                } else {
                    cacheIndex = DEBUG_IMAGE;
                    cacheFile = new File(cacheDir, DEBUG_IMAGE + ".png");
                    tempTime = System.currentTimeMillis();
                }

                // 读取图片像素
                BufferedImage image = ImageIO.read(cacheFile);
                final int screenWidth = image.getWidth();
                final int screenHeight = image.getHeight();
                debugInfo.append("屏幕参数 ").append(screenWidth).append("x")
                        .append(screenHeight).append("\n");
                if (screenWidth == 0 || screenHeight == 0) {
                    System.out.println(cacheIndex + ". 截屏图片读取错误");
                    Thread.sleep(5000);
                    continue;
                }
                Pixel[][] pixels = new Pixel[screenHeight][screenWidth];
                for (int i = 0; i < screenHeight; i++) {
                    for (int j = 0; j < screenWidth; j++) {
                        pixels[i][j] = new Pixel();
                        pixels[i][j].color = image.getRGB(j, i);
                        pixels[i][j].x = j;
                        pixels[i][j].y = i;

                    }
                }

                final int TOP_BORDER = HelperUtil.transH(screenHeight, HelperUtil.adjustTopBorder(screenWidth,screenHeight));
                final int BOTTOM_BORDER = HelperUtil.transH(screenHeight, HelperUtil.adjustBottomBorder(screenWidth,screenHeight));
                final int LEFT_BORDER = HelperUtil.transW(screenWidth, Constants.LEFT_BORDER);
                final int RIGHT_BORDER = HelperUtil.transW(screenWidth, Constants.RIGHT_BORDER);

                final int PIECE_WIDTH_PIXELS = HelperUtil.transW(screenWidth, Constants.PIECE_WIDTH_PIXELS);
                final int PIECE_TOP_PIXELS = HelperUtil.transW(screenWidth, Constants.PIECE_TOP_PIXELS);
                final int PIECE_BOTTOM_CENTER_SHIFT = HelperUtil.transW(screenWidth, Constants.PIECE_BOTTOM_CENTER_SHIFT);

                final int DES_MIN_SHIFT = HelperUtil.transW(screenWidth, Constants.DES_MIN_SHIFT);

				/* 计算棋子位置 */
                Pixel piece = new Pixel();
                for (int i = TOP_BORDER; i < screenHeight - BOTTOM_BORDER; i++) {
                    int startX = 0;
                    int endX = 0;
                    for (int j = LEFT_BORDER; j < screenWidth - RIGHT_BORDER; j++) {
                        int red = Color.red(pixels[i][j].color);
                        int green = Color.green(pixels[i][j].color);
                        int blue = Color.blue(pixels[i][j].color);
                        if (50 < red && red < 55 && 50 < green && green < 55
                                && 55 < blue && blue < 65) {// 棋子顶部颜色
                            // 如果侦测到棋子相似颜色，记录下开始点
                            if (startX == 0) {
                                startX = j;
                                endX = 0;
                            }
                        } else if (endX == 0) {
                            // 记录下结束点
                            endX = j;

                            if (endX - startX < PIECE_TOP_PIXELS) {
                                // 规避井盖的BUG，像素点不够长，则重新计算
                                startX = 0;
                                endX = 0;
                            }
                        }
                        if (50 < red && red < 60 && 55 < green && green < 65
                                && 95 < blue && blue < 105) {// 棋子底部的颜色
                            // 最后探测到的颜色就是棋子的底部像素
                            piece.y = i;
                        }
                    }
                    if (startX != 0 && piece.x == 0) {
                        piece.x = (startX + endX) / 2;
                    }
                }
                if (piece.x == 0 || piece.y == 0) {
                    System.out.println(cacheIndex + ". 未找到棋子坐标");
                    final int PRESS_Y = HelperUtil.transH(screenHeight, Constants.RESTART_PRESS_Y);
                    final int PRESS_X = HelperUtil.transW(screenWidth, Constants.RESTART_PRESS_X);
                    HelperUtil.execute(Constants.ADB_PATH.getAbsolutePath(), "shell", "input", "swipe", PRESS_X + "", PRESS_Y + "", PRESS_X + "", PRESS_Y + "", 100 + "");
                    Thread.sleep(5000);
                    continue;
                }

                // 棋子纵坐标从底部边缘调整到底部中心
                piece.y -= PIECE_BOTTOM_CENTER_SHIFT;

                System.out.print(cacheIndex + ". 棋子坐标点[" + piece.x + ", " + piece.y + "], ");

                // 计算目标位置
                Pixel des = new Pixel();

                Pixel firstPixcel = null;// 目标落点最顶部首个像素位置

                // 双重循环寻找落点第一个像素
                for (int i = TOP_BORDER; i < screenHeight - BOTTOM_BORDER; i++) {
                    for (int j = LEFT_BORDER; j < screenWidth - RIGHT_BORDER; j++) {
                        if (piece.x - PIECE_WIDTH_PIXELS / 2 < j && j < piece.x + PIECE_WIDTH_PIXELS / 2) {
                            // 忽略和棋子处于同一垂直线的这些像素
                            continue;
                        }
                        if (!Color.compareColor(pixels[i][j].color, pixels[i][0].color, 12)) {
                            // 发现色差，记录下落点顶部第一个像素
                            firstPixcel = pixels[i][j];
                            break;
                        }
                    }
                    if (firstPixcel != null) {
                        break;
                    }
                }
                if (firstPixcel == null) {
                    System.out.println(cacheIndex + ". 未找到目标落点坐标");
                    Thread.sleep(5000);
                    continue;
                }
                debugInfo.append("落点首个像素 firstPixel").append(firstPixcel)
                        .append("\n");

                // 计算目标点类型

                Pixel whitePointCenter = findWhitePointCenter(pixels, firstPixcel);
                Pixel puerCenter = findPuerCenter(pixels, firstPixcel);

                DesType desType = DesTypeChecker.getDesType(pixels, firstPixcel, puerCenter);

                String hasWhitePoint;
                if (whitePointCenter != null) {
                    // 如果有白点，那目标落点为白点中心
                    des = whitePointCenter;
                    hasWhitePoint = "有靶点";
                } else {
                    hasWhitePoint = "无靶点";
                    if (puerCenter != null) {
                        des = puerCenter;
                    } else {
                        // 如果是其他类型且没有靶点，通过斜率计算落点
                        int count = 0;
                        for (int i = firstPixcel.x; i < screenWidth - RIGHT_BORDER; i++) {
                            if (Color.compareColor(pixels[firstPixcel.y][i].color, pixels[firstPixcel.y][firstPixcel.x].color, 10)) {
                                ++count;
                            } else {
                                break;
                            }
                        }

                        des.x = firstPixcel.x + count / 2;
                        float k;// 斜率
                        if ((firstPixcel.y - piece.y) * 1.0f / (firstPixcel.x - piece.x) < 0) {
                            k = Constants.K1;
                        } else {
                            k = Constants.K2;
                        }
                        des.y = Math.round(k * (des.x - piece.x) + piece.y);
                        if (des.y < firstPixcel.y + DES_MIN_SHIFT) {
                            des.y = firstPixcel.y + DES_MIN_SHIFT;
                        }
                    }
                }

				/* 计算距离 */
                double distance = HelperUtil.calculateDistance(piece, des);
                System.out.print("目标落点[" + des.x + ", " + des.y + "], " + desType.getName() + ", " + hasWhitePoint + ", 距离" + Math.round(distance) + "px, ");
                calculateTime = System.currentTimeMillis() - tempTime;
                tempTime += calculateTime;
                if (!DEBUG) {
                    // 执行跳跃
                    long pressTime = (long) (distance * jumpParam);
                    System.out.println("模拟按压" + pressTime + "ms.");
                    final int PRESS_Y = HelperUtil.transH(screenHeight, screenHeight / 2 - 250 + (int) (Math.random() * 500));
                    final int PRESS_X = HelperUtil.transW(screenWidth, screenWidth / 2 - 150 + (int) (Math.random() * 300));
                    HelperUtil.execute(Constants.ADB_PATH.getAbsolutePath(), "shell", "input", "swipe", PRESS_X + "", PRESS_Y + "", PRESS_X + "", PRESS_Y + "", pressTime + "");
                    exeJumpTime = System.currentTimeMillis() - tempTime;
                    tempTime += exeJumpTime;
                    // 保存带标记的图像
                    Graphics graphics = image.getGraphics();
                    graphics.setColor(java.awt.Color.RED);
                    graphics.drawLine(des.x, des.y, piece.x, piece.y);
                    graphics.setFont(new Font("宋体", Font.BOLD, HelperUtil.transW(screenWidth, 60)));
                    graphics.drawString(desType.getName() + "，" + hasWhitePoint, HelperUtil.transW(screenWidth, 10), HelperUtil.transW(screenWidth, 70));
                    graphics.drawString("跳跃距离" + Math.round(distance) + "px", HelperUtil.transW(screenWidth, 10), HelperUtil.transW(screenWidth, 135));
                    graphics.drawString("模拟按压" + pressTime + "ms", HelperUtil.transW(screenWidth, 10), HelperUtil.transW(screenWidth, 200));
                    ImageIO.write(image, "png", new File(markDir, cacheIndex + "_mark.png"));
                    recacheTime = System.currentTimeMillis() - tempTime;
                } else {
                    System.out.println("DEBUG");
                    System.out.print(debugInfo);
                }

                System.out.print("截图耗时" + screencapTime + "ms, ");
                System.out.print("下载耗时" + downloadTime + "ms, ");
                System.out.print("缓存耗时" + cacheTime + "ms, ");
                System.out.print("计算耗时" + calculateTime + "ms, ");
                System.out.print("跳跃耗时" + exeJumpTime + "ms, ");
                System.out.print("标记耗时" + recacheTime + "ms, ");
                System.out.println("总耗时"
                        + (System.currentTimeMillis() - startTime) + "ms\n");

				/* 停留一会，保证棋子落稳 */
                if (desType == DesType.COVER || desType == DesType.CUBE
                        || desType == DesType.SHOP || desType == DesType.DISC) {
                    // 特殊类型的落点，停留时间长了会加分
                    Thread.sleep(2000);
                } else {
                    Thread.sleep(1500);
                }
            } catch (Exception e) {
                e.printStackTrace();
            }
        }
    }

    /**
     * 获取纯平像素点中心，如果检测到不是纯平面，则返回null
     */
    public Pixel findPuerCenter(Pixel[][] pixels, Pixel firstPixcel) {

        Pixel[] vertexs = HelperUtil.findVertexs(pixels, firstPixcel);
        Pixel puerLeft = vertexs[0];
        Pixel puerTop = vertexs[1];
        Pixel puerRight = vertexs[2];
        Pixel puerBottom = vertexs[3];
        Pixel puerCenter;

        debugInfo.append("PuerBorder[")
                .append(puerLeft.x).append(", ")
                .append(puerTop.y).append(", ")
                .append(puerRight.x).append(", ")
                .append(puerBottom.y).append("]\n");
        int screenWidth = pixels[0].length;
        int screenHeight = pixels.length;
        int puerWidth = puerRight.x - puerLeft.x + 1;
        int puerHeight = puerBottom.y - puerTop.y + 1;
        int distanceMin = HelperUtil.transW(screenWidth, 30);
        if (HelperUtil.transW(screenWidth, 110) <= puerWidth
                && puerWidth <= HelperUtil.transW(screenWidth, 610)
                && HelperUtil.transW(screenWidth, 64) <= puerHeight
                && puerHeight <= HelperUtil.transW(screenWidth, 350)
                && HelperUtil.calculateDistance(puerTop, puerLeft) > distanceMin
                && HelperUtil.calculateDistance(puerTop, puerRight) > distanceMin
                && HelperUtil.calculateDistance(puerBottom, puerLeft) > distanceMin
                && HelperUtil.calculateDistance(puerBottom, puerRight) > distanceMin) {
            puerCenter = pixels[(puerTop.y + puerBottom.y) / 2][(puerLeft.x + puerRight.x) / 2];
        } else {
            puerCenter = null;
        }
        return puerCenter;
    }

    /**
     * 获取靶点(落点中心的白点)中心，如果检测到没有靶点，则返回null
     */
    private Pixel findWhitePointCenter(Pixel[][] pixels, Pixel firstPixcel) {
        int screenWidth = pixels[0].length;
        int screenHeight = pixels.length;
        // 尝试寻找白点
        Pixel whitePointBase = null;
        final int WHITE_POINT_COLOR = 0xf5f5f5;
        int widthShift = HelperUtil.transW(screenWidth, 15);
        int heightShift = HelperUtil.transW(screenWidth, 200);
        for (int i = firstPixcel.y; i < firstPixcel.y + heightShift; i++) {
            for (int j = firstPixcel.x - widthShift; j < firstPixcel.x + widthShift; j++) {
                if (Color.compareColor(pixels[i][j].color, WHITE_POINT_COLOR)) {
                    // 发现白点相似颜色
                    whitePointBase = pixels[i][j];
                    break;
                }
            }
            if (whitePointBase != null) {
                break;
            }
        }
        if (whitePointBase == null) {
            return null;
        }
        if (DesTypeChecker.checkBumf(pixels, firstPixcel)) {
            // 如果是卫生纸,修正白色点的基准
            whitePointBase = pixels[firstPixcel.y + HelperUtil.transW(screenWidth, 62)][firstPixcel.x];
        }

        Pixel[] vertexs = HelperUtil.findVertexs(pixels, whitePointBase);
        Pixel whitePointLeft = vertexs[0];
        Pixel whitePointTop = vertexs[1];
        Pixel whitePointRight = vertexs[2];
        Pixel whitePointBottom = vertexs[3];
        Pixel whitePointCenter;

        debugInfo.append("WhitePointBorder[").append(whitePointLeft.x)
                .append(", ").append(whitePointTop.y).append(", ")
                .append(whitePointRight.x).append(", ")
                .append(whitePointBottom.y).append("]\n");
        int whitePointWidth = whitePointRight.x - whitePointLeft.x + 1;
        int whitePointHeight = whitePointBottom.y - whitePointTop.y + 1;
        if (HelperUtil.transW(screenWidth, Constants.WHITE_POINT_MIM_WIDTH) <= whitePointWidth
                && whitePointWidth <= HelperUtil.transW(screenWidth, Constants.WHITE_POINT_MAX_WIDTH)
                && HelperUtil.transW(screenWidth, Constants.WHITE_POINT_MIM_HEIGHT) <= whitePointHeight
                && whitePointHeight <= HelperUtil.transW(screenWidth, Constants.WHITE_POINT_MAX_HEIGHT)) {
            whitePointCenter = pixels[(whitePointTop.y + whitePointBottom.y) / 2][(whitePointLeft.x + whitePointRight.x) / 2];
        } else {
            whitePointCenter = null;
        }
        return whitePointCenter;
    }

}
